lib/charms/mongodb/v0/shards_interface.py

```diff
@@ -0,0 +1,211 @@
+        """Updates a set of key-value pairs in the relation.
+
+        This function writes in the application data bag, therefore, only the leader unit can call
+        it.
+
+        Args:
+            relation_id: the identifier for a particular relation.
+            data: dict containing the key-value pairs
+                that should be updated in the relation.
+        """
+        if self.charm.unit.is_leader():
+            relation = self.charm.model.get_relation(self.relation_name, relation_id)
+            if relation:
+                relation.data[self.charm.model.app].update(data)
+
+
+class ConfigServerRequirer(Object):
+    """Manage relations between the config server and the shard, on the shard's side."""
+
+    def __init__(
+        self, charm: CharmBase, relation_name: str = Config.Relations.SHARDING_RELATIONS_NAME
+    ) -> None:
+        """Constructor for ShardingProvider object."""
+        self.relation_name = relation_name
+        self.charm = charm
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+
+        # TODO Future PR, enable shard drainage by observing relation departed events
+
+    def _on_relation_changed(self, event):
+        """Retrieves secrets from config-server and updates them within the shard."""
+        if self.charm.is_role(Config.Role.REPLICATION):
+            self.unit.status = BlockedStatus("role replication does not support sharding")
+            logger.error("sharding interface not supported with config role=replication")
+            return
+
+        if not self.charm.is_role(Config.Role.SHARD):
+            logger.info(
+                "skipping relation changed event ShardingProvider is only be executed by shards"
+            )
+            return
+
+        if not self.charm.db_initialised:
+            event.defer()
+
+        # shards rely on the config server for secrets
+        relation_data = event.relation.data[event.app]
+        self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        try:
+            self.update_operator_password(new_password=relation_data.get(OPERATOR_PASSWORD_KEY))
+        except RetryError:
+            self.charm.unit.status = BlockedStatus("Shard not added to config-server")
+            return
+
+        self.charm.unit.status = MaintenanceStatus("Adding shard to config-server")
+
+        # TODO future PR, leader unit verifies shard was added to cluster
+        if not self.charm.unit.is_leader():
+            return
+
+    def update_operator_password(self, new_password: str) -> None:
+        """Updates the password for the operator user.
+
+        Raises:
+            RetryError
+        """
+        current_password = (
+            self.charm.get_secret(
+                Config.Relations.APP_SCOPE, MongoDBUser.get_password_key_name_for_user("operator")
```

Comment:
getting secret can be slow (up to 2 seconds), can we add the following block before getting secret
```
if not new_password or not self.charm.unit.is_leader():
    return
```

Reply:
Ah, good idea, thank you Mykola

Reply:
we cache secrest in moment of the first retrieval till the end of event